keps/sig-scheduling/2249-pod-affinity-namespace-selector/README.md

```diff
@@ -0,0 +1,663 @@
+cases of all pods having affinity/anti-affinity constraints to/against each other.
+Those benchmarks show that the scheduler is able to achieve maximum throughput 
+(i.e., the api-server qps limit).
+
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+ Add `NamespaceSelector` field to `PodAffinityTerm`:
+
+```go
+type PodAffinityTerm struct {
+    // A label query over the set of namespaces for which the term applies to.
+    // The term is applied to the union of the namespaces selected by this field
+    // and the ones listed in the `Namespaces` field.
+	NamespaceSelector *metav1.LabelSelector
+}
+```
+
+As indicated in the comment, the scheduler will consider the union of the namespaces
+specified in the existing `Namespaces` field and the ones selected by the new 
+`NamespaceSelector` field. `NamespaceSelector`  is ignored when set to nil.
+
+The feature will be guarded by a new feature flag. When disabled, the api-server 
+will silently drop the `NamespaceSelector` field if set, moreover kube-scheduler 
+will ignore the field and continue to behave as before.
+
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+### Test Plan
+
+- **Unit Tests**: core changes are covered by unit tests.
+- **Integration Tests**: core changes covered by integration tests.
+- **Benchmark Tests**: The performance should be as close as possible to the case where all pods are created in one namespace.
+
+<!--
+**Note:** *Not required until targeted at a release.*
+
+Consider the following in developing a test plan for this enhancement:
+- Will there be e2e and integration tests, in addition to unit tests?
+- How will it be tested in isolation vs with other components?
+
+No need to outline all of the test cases, just the general strategy. Anything
+that would count as tricky in the implementation, and anything particularly
+challenging to test, should be called out.
+
+All code is expected to have adequate tests (eventually with coverage
+expectations). Please adhere to the [Kubernetes testing guidelines][testing-guidelines]
+when drafting this test plan.
+
+[testing-guidelines]: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
+-->
+
+### Graduation Criteria
+
+#### Alpha -> Beta Graduation
+
+    - No complaints regarding performance.
+
+#### Beta -> GA Graduation
+    
+    - Still no complaints regarding performance.
```

Comment:
the bar for this should be higher... we should actively prove there are not performance problems with large numbers of namespaces and pods selected by affinity/anti-affinity rules, and the scheduler can keep up with pod creation/deletion and namespace label changes